oak_crypto/src/aead.rs

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{schema::AeadEncryptedMessage, MessageWithAssociatedData};
+use aes_gcm::{aead::AeadInPlace, Aes256Gcm, Key, KeyInit, Nonce};
+use anyhow::anyhow;
+
+/// AES-256-GCM ID.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-authenticated-encryption-wi>
+pub(crate) const AEAD_ID: u16 = 0x0002;
+/// Represents `N_k` from RFC9180.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-cryptographic-dependencies>
+pub(crate) const AEAD_ALGORITHM_KEY_SIZE_BYTES: usize = 32;
+/// Represents `N_n` from RFC9180.
+/// <https://www.rfc-editor.org/rfc/rfc9180.html#name-cryptographic-dependencies>
+pub(crate) const AEAD_NONCE_SIZE_BYTES: usize = 12;
+/// Convenience type for representing an AEAD key.
+pub(crate) type AeadKey = [u8; AEAD_ALGORITHM_KEY_SIZE_BYTES];
+/// Convenience type for representing an AEAD nonce.
+pub(crate) type AeadNonce = [u8; AEAD_NONCE_SIZE_BYTES];
+
+/// Implementation of Authenticated Encryption with Associated Data (AEAD).
+/// <https://datatracker.ietf.org/doc/html/rfc5116>
+pub(crate) struct Aead {}
```

Comment:
For a lot of these the struct contains no state. Is there any point to having them? Are you planning to store some state in the struct later on?

If not, the encrypt etc functions could be just free functions.

Reply:
+1 not sure why all the empty structs; if these were functions, they would already be namespaced based on the mod, which seems enough